### PR TITLE
net: openthread: rpc: CoAP APIs serialization (part 1)

### PR DESCRIPTION
**Pull request by Damian-Nordic**
Add serialization of the following OpenThread APIs:
  - otCoapNewMessage
  - otCoapMessageInit
  - otCoapMessageInitResponse
  - otCoapMessageAppendUriPathOptions
  - otCoapMessageSetPayloadMarker
  - otCoapMessageGetType
  - otCoapMessageGetCode
  - otCoapMessageGetMessageId
  - otCoapMessageGetTokenLength
  - otCoapMessageGetToken
  - otCoapStart
  - otCoapStop
  - otCoapAddResource
  - otCoapRemoveResource
  - otCoapSetDefaultHandler
    
The implementation of CoAP handlers registered with otCoapAddResource and otCoapSetDefaultHandler is incomplete and will be reworked in a subsequent PR.

### CODE REVIEW

**Review by bpienk — APPROVED** [review]

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+	return error;
+}
+
+otError otCoapMessageAppendUriPathOptions(otMessage *aMessage, const char *aUriPath)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	size_t cbor_buffer_size = 0;
+	otError error;
+
+	cbor_buffer_size += 1 + sizeof(uintptr_t); /* aMessage */
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:82]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+}
+
+otError otCoapMessageInitResponse(otMessage *aResponse, const otMessage *aRequest, otCoapType aType,
+				  otCoapCode aCode)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	size_t cbor_buffer_size = 0;
+	otError error;
+
+	cbor_buffer_size += 1 + sizeof(uintptr_t); /* aResponse */
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:59]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+	cbor_buffer_size += 1 + sizeof(uintptr_t); /* aRequest */
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:60]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+	return (otMessage *)message_rep;
+}
+
+void otCoapMessageInit(otMessage *aMessage, otCoapType aType, otCoapCode aCode)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	size_t cbor_buffer_size = 0;
+
+	cbor_buffer_size += 1 + sizeof(uintptr_t); /* aMessage */
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:39]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+	return error;
+}
+
+otError otCoapMessageSetPayloadMarker(otMessage *aMessage)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	otError error;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 1 + sizeof(uintptr_t));
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:100]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)aMessage);
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:101]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+
+	return error;
+}
+
+otCoapType otCoapMessageGetType(const otMessage *aMessage)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	otCoapType type = 0;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 1 + sizeof(uintptr_t));
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:114]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)aMessage);
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:115]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+	return type;
+}
+
+otCoapCode otCoapMessageGetCode(const otMessage *aMessage)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	otCoapCode code = 0;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 1 + sizeof(uintptr_t));
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)aMessage);
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:130]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+
+	return type;
+}
+
+otCoapCode otCoapMessageGetCode(const otMessage *aMessage)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	otCoapCode code = 0;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 1 + sizeof(uintptr_t));
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:129]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+
+	return code;
+}
+
+uint16_t otCoapMessageGetMessageId(const otMessage *aMessage)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	uint16_t id = 0;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 1 + sizeof(uintptr_t));
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:144]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)aMessage);
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:145]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+
+	return id;
+}
+
+uint8_t otCoapMessageGetTokenLength(const otMessage *aMessage)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	uint8_t token_length = 0;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 1 + sizeof(uintptr_t));
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:158]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)aMessage);
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:159]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+
+	return token_length;
+}
+
+const uint8_t *otCoapMessageGetToken(const otMessage *aMessage)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	static uint8_t token[OT_COAP_MAX_TOKEN_LENGTH];
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 1 + sizeof(uintptr_t));
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:172]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+	nrf_rpc_encode_uint(&ctx, (uintptr_t)aMessage);
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:173]: Please use `ot_msg_key` instedo of `uintptr_t`.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c (hunk 1 above)
Comment (Damian-Nordic) [reply subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:82]: I assumed that this is implementation detail of the server and the client should assume this is just `otMessage*` being sent but I can change it.

File: subsys/net/openthread/rpc/client/ot_rpc_coap_client.c [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+	nrf_rpc_encode_str(&ctx, aResource->mUriPath, -1);
+
+	nrf_rpc_cbor_cmd_no_err(&ot_group, OT_RPC_CMD_COAP_REMOVE_RESOURCE, &ctx,
+				nrf_rpc_rsp_decode_void, NULL);
+}
+
+static void ot_rpc_cmd_coap_resource_handler(const struct nrf_rpc_group *group,
+					     struct nrf_rpc_cbor_ctx *ctx, void *handler_data)
+{
+	uintptr_t message_rep = nrf_rpc_decode_uint(ctx);
```
Comment (canisLupus1313) [inline subsys/net/openthread/rpc/client/ot_rpc_coap_client.c:268]: This looks like an unused variable. Why do You need it?

**Review by canisLupus1313 — APPROVED** [review]

**Pull request merged**